### ISSUE TITLE
Request: smaller dependency surface on Transformers model types?

### ISSUE THREAD

**Issue by andysalerno** [guidance-ai/guidance#408]
Hi, this is a very vague feature request :D

There are some parts of the code (in the main branch, I haven't looked at the upcoming branches that were mentioned in #395 ) that depend on some private/internal functionality of Transformers types.  Mostly around the cache and logits history.

The problem with this is, it's difficult to create wrappers for other model providers, like GPTQ, AQW, llamacpp etc etc because you have to replicate some pretty specific private code from the Transformer types that Guidance expects to see.

So my request is this - is it possible to define a smaller interface/surface, not specific to Transformers (or at least not relying on anything internal to Transformers), that can be used to host models from other sources?

No idea how difficult this is or if it's realistic, just something that I would personally love to see.

**Comment by freckletonj** [comment]
I think `guidance` is effectively dead now. The maintainers are nearly radio silent, and although i don't love `lmql`, it's the best I'm aware of right now, so, maybe you'll get more mileage there? I had gptq stuff working with them, and I think they claim compatibility with llamacpp.

That said, if this is something you go after, here are a couple relevant `transformers` PRs that are changing caching related things:

https://github.com/huggingface/transformers/pull/26681

https://github.com/huggingface/transformers/pull/25086

**Comment by freckletonj** [comment]
@paucodeici I agree with everything you've said about lmql. It's a pain to use and debug. I still haven't found anything better yet though. Guidance *was* way better, but alas, it seems pretty dead.

Have you tried the `pythonic` branch?

**Comment by paucodeici** [comment]
The way it will be handled seem promising when you look the commit (at
least it will be way more pythonic but i feel they are still not sure how
to expose the things to be "pythonic").
From my last test it was not working for my use case but this is a work in
progress branch so it has to be expected things can be broken.

Le mar. 7 nov. 2023 à 21:26, Josh Freckleton ***@***.***> a
écrit :

**Comment by marcotcr** [comment]
Please check out our current attempt in the [new release](https://github.com/guidance-ai/guidance/discussions/429), where we integrate llama.cpp as well.
